tests/testthat/test-equal.R

```diff
@@ -232,6 +239,16 @@ test_that("not equal if attributes not equal", {
   expect_false(obj_equal(x1, x2))
 })
 
+test_that("can compare expressions", {
+  expect_true(obj_equal(expression(x), expression(x)))
+  expect_false(obj_equal(expression(x), expression(y)))
+})
+
+test_that("`na_equal` works with expressions", {
+  expect_true(obj_equal(expression(NA), expression(NA)))
```

Comment:
This is the only thing I have qualms about (and it probably isn't that important), but I'd think that the `NA` of an expression would be `NULL`, like lists.

Reply:
Oh I think i have the test wrong, but it works just like lists

``` r
obj_equal <- vctrs:::obj_equal

obj_equal(expression(NA), expression(NA), na_equal = TRUE)
#> [1] TRUE
obj_equal(list(NA), list(NA), na_equal = TRUE)
#> [1] TRUE

obj_equal(expression(NA), expression(NA), na_equal = FALSE)
#> [1] NA
obj_equal(list(NA), list(NA), na_equal = FALSE)
#> [1] NA

obj_equal(list(NULL), list(NULL), na_equal = TRUE)
#> [1] TRUE
obj_equal(expression(NULL), expression(NULL), na_equal = TRUE)
#> [1] TRUE

obj_equal(list(NULL), list(NULL), na_equal = FALSE)
#> [1] NA
obj_equal(expression(NULL), expression(NULL), na_equal = FALSE)
#> [1] NA
```

<sup>Created on 2019-11-13 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0.9000)</sup>

Reply:
Oh the `NA` result works because you're recursing down the tree.

Reply:
Now I am surprised to see `NA` propagation for scalar equality. I don't think `obj_equal()` should have an `na_equal` parameter at all because NA propagation is a property of vector operations.